### PR TITLE
Validate during the build step that no non-browser APIs are used.

### LINKED ISSUE #1531: `inlang module build` should fail if node (or other non-browser) APIs are used
## Problem 

Critical bugs occur in production like https://github.com/inlang/monorepo/issues/1530 because `inlang module build` is not failing when node (or other non-browser) APIs are used. 

## Proposal 

Validate during the build step that no non-browser APIs are used. 

### How

Add an esbuild plugin to the `inlang module build` command that throws if node imports are encountered. 

https://github.com/inlang/monorepo/blob/d7710fcf2427aad5f53125f7f52375fa468e91e9/inlang/source-code/cli/src/commands/module/build/command.ts#L28-L41

## Additional information

- node APIs can still be be polyfilled. validation ensures that they are actually polyfilled (the reason why https://github.com/inlang/monorepo/issues/1530 was caused)
- polyfilling by default (what we used to do) is setting up the ecosystem to be non-web compliant aka a no-go

### PR DESCRIPTION
**Pull request by Min2who**
/claim #1531
fixed #1531

### CODE REVIEW

**Comment by Min2who** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Min2who** [comment]
recheck

**Comment by Min2who** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by Min2who** [comment]
can you review the PR once

File: inlang/source-code/cli/src/commands/module/build/command.ts [hunk 1]
```diff
@@ -8,6 +8,22 @@ import { moduleBuildOptions } from "./moduleBuildOptions.js"
  */
 import { context } from "esbuild-wasm"
 
+// Define the forbidden Node.js imports
+const forbiddenNodeImports = ['fs', 'path', 'os', 'net']
+
+// Create an esbuild plugin to check Node.js imports
+const nodeAPICheckerPlugin = {
+	name: 'node-api-checker',
+	setup(build) {
```
Comment (flornkm) [inline inlang/source-code/cli/src/commands/module/build/command.ts:17]: CI is failing here as no type is set

File: inlang/source-code/cli/src/commands/module/build/command.ts [hunk 2]
```diff
@@ -8,6 +8,22 @@ import { moduleBuildOptions } from "./moduleBuildOptions.js"
...
+	  build.onResolve({ filter: /./ }, (args) => {
```
Comment (flornkm) [inline inlang/source-code/cli/src/commands/module/build/command.ts:18]: And here too, see the CI pipeline:

Screenshot 2023-11-16 at 12 30 34 PM

**Review by flornkm — COMMENTED** [review]
Some types are missing and results in the CI failing

File: inlang/source-code/cli/src/commands/module/build/command.ts (hunk 2 above)
Comment (felixhaeberle) [reply inlang/source-code/cli/src/commands/module/build/command.ts:18]: Link to CI: https://github.com/inlang/monorepo/actions/runs/6886826790/job/18739877557

**Comment by samuelstroschein** [comment]
Hi @Min2who, 

Can you add tests for the code and add all node api's? there are more than the 4 you used.

**Comment by Min2who** [comment]
> Hi @Min2who,
> 
> Can you add tests for the code and add all node api's? there are more than the 4 you used.

@samuelstroschein will it be okay if i complete the test by tommorow i am occupied a bit

**Comment by samuelstroschein** [comment]
@Min2who yes :)

**Comment by Min2who** [comment]
> Hi @Min2who,
> 
> Can you add tests for the code and add all node api's? there are more than the 4 you used.

@samuelstroschein i have added the test as you have mention

File: inlang/source-code/cli/src/commands/module/build/buildCommandAction.test.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
+import { it, expect } from 'vitest';
+import { buildCommandAction } from './command.js';
+import type { BuildActionArgs } from './command.js';
+
+// List all Node.js APIs that should be forbidden
+const forbiddenNodeImports: string[] = [
+  'fs',
+  'path',
+  'os',
+  'net',
```
Comment (samuelstroschein) [inline inlang/source-code/cli/src/commands/module/build/buildCommandAction.test.ts:10]: can you add `node:fs`, `node:path` etc as well?

File: inlang/source-code/cli/src/commands/module/build/buildCommandAction.test.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  };
+
+  for (const forbiddenImport of forbiddenNodeImports) {
+    await it(`Should detect forbidden Node.js import: ${forbiddenImport}`, async () => {
+      const args: BuildActionArgs = {
+        entry: 'testEntry.js', // Replace with an actual entry file path
+        outdir: './dist',
+        watch: false,
+      };
+
```
Comment (samuelstroschein) [inline inlang/source-code/cli/src/commands/module/build/buildCommandAction.test.ts:30]: How does this test work? 

I would expect multiple test files that import node APIs. I see no `testEntry.js` file however.

**Comment by Min2who** [comment]
Can you please explain to me what test are you exactly expecting

**Comment by samuelstroschein** [comment]
> Can you please explain to me what test are you exactly expecting

That test the behavior of importing node APIs in a JavaScript file. 

```ts

const mockFile = `
import fs from "fs"

console.log(fs)
`

// run test
```

**Comment by Min2who** [comment]
@samuelstroschein i have reattempt writing the test can you review it once

**Comment by samuelstroschein** [comment]
Closing this because it overwhelms me to review the PR and express what I want. It seems easier to implement this myself. 

In particular, I don't understand how the test pass/fail without mocking the filesystem that esbuild reads from. I fear that this test/PR is not testing the actual behavior.

**Pull request closed without merging**